src/Middleware/ImplicitHeadMiddleware.php

```diff
@@ -47,24 +50,25 @@ class ImplicitHeadMiddleware implements MiddlewareInterface
     public const FORWARDED_HTTP_METHOD_ATTRIBUTE = 'forwarded_http_method';
 
     /**
-     * @var ResponseInterface
+     * @var callable
      */
-    private $response;
+    private $responseFactory;
 
     /**
      * @var callable
      */
     private $streamFactory;
 
     /**
-     * @param ResponseInterface $response Response prototype to return for
+     * @param callable $responseFactory A factory capable of returning an
+     *     empty ResponseInterface instance to return to return for
      *     implicit HEAD requests.
      * @param callable $streamFactory A factory capable of returning an empty
      *     StreamInterface instance to inject in a response.
      */
-    public function __construct(ResponseInterface $response, callable $streamFactory)
+    public function __construct(callable $responseFactory, callable $streamFactory)
     {
-        $this->response = $response;
+        $this->responseFactory = $responseFactory;
```

Comment:
what if we do it this way to enforce we will get what we expect?
```
$this->responseFactory = function() use ($responseFactory) : ResponseInterface {
    return $responseFactory();
};
```

Reply:
Or version with parameters:
```php
$this->responseFactory = function(...$params) use ($responseFactory) : ResponseInterface {
    return $responseFactory(...$params);
};
```

Reply:
@Ocramius what do you think about this? Is it over the top?

Reply:
First, as noted: we won't accept parameters, or, if we do, they will match those of PSR-17. I'm perfectly fine with using `($this->responseFactory)()` to produce the response.

Alternately, we make the factory an object with named methods. My only concern with doing that is that if PSR-17 renames any methods before it is finalized, we end up with BC breaks and/or compat shims, which will be problematic. Callables will always work, however, and we can even pass a reference to a PSR-17 factory method (`return [$factoryClass, 'createResponse]`) or a closure around it (`return function () use ($factory) { return $factory->createResponse(); };`).